gpflow/base.py

```diff
@@ -10,38 +10,42 @@
 
 from .config import default_float
 
+if TYPE_CHECKING:
+    from IPython.lib import pretty
+
 DType = Union[np.dtype, tf.DType]
-VariableData = Union[List, Tuple, np.ndarray, int, float]
+VariableData = Union[List, Tuple, np.ndarray, int, float]  # deprecated
 Transform = Union[tfp.bijectors.Bijector]
 Prior = Union[tfp.distributions.Distribution]
+UNKNOWN = Any
 
+S = TypeVar("S")
+_NestedSeq = Union[
+    Sequence[S],
+    Sequence[Sequence[S]],
+    Sequence[Sequence[Sequence[S]]],
+    Sequence[Sequence[Sequence[Sequence[Any]]]],  # catch all for rank > 3
+]
+_NativeScalar = Union[int, float]
 
-def _IS_PARAMETER(o):
-    return isinstance(o, Parameter)
 
+_ArrayOrScalar = Union[_NativeScalar, _NestedSeq[_NativeScalar]]
+# represents an `int`, `float`, or nested `Sequence` of `int` or `float`, such as
+#
+# 1, 1.1, [1.1], or [[1.1], [1, 1.1]]
+#
+# The ranks must be consistent, but the lengths are unconstrained
 
-def _IS_TRAINABLE_PARAMETER(o):
-    return _IS_PARAMETER(o) and o.trainable
 
+_TensorTypeExternal = Union[tf.Tensor, tf.Variable, np.ndarray]
```

Comment:
Can also be int and float for scalar tensors

Reply:
keenness get the better of you ti?

Reply:
this is a good point. I'm just trying to think if this is right generally

Reply:
re :/ i thought planned to hand this PR over to eric since your week was up

Reply:
wrt ints and floats, they don't have `__matmul__` i.e. you can't do `1 @ 1`. I think this means they can't always be treated in quite the same way as `Tensor`s